core/src/main/java/org/apache/iceberg/rest/auth/GenericAuthSessionCache.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest.auth;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Ticker;
+import java.time.Duration;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.Uninterruptibles;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A cache for {@link AuthSession} instances. */
+public class GenericAuthSessionCache<K, V extends AuthSession> implements AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GenericAuthSessionCache.class);
+
+  private final Duration sessionTimeout;
+  private final Executor executor;
+  private final Ticker ticker;
+
+  private volatile Cache<K, V> sessionCache;
+
+  /**
+   * Creates a new cache with the given session timeout, and with default executor and default
+   * ticker for eviction tasks.
+   *
+   * @param name a distinctive name for the cache.
+   * @param sessionTimeout the session timeout. Sessions will become eligible for eviction after
+   *     this duration of inactivity.
+   */
+  public GenericAuthSessionCache(String name, Duration sessionTimeout) {
+    this(
+        sessionTimeout,
+        ThreadPools.newExitingWorkerPool(name + "-auth-session-evict", 1),
```

Comment:
should this be using the new auth pool once it's merged?

Reply:
@danielcweeks suggested a dedicated pool here:

https://github.com/apache/iceberg/pull/11844#discussion_r1917315859

I'm fine either way.